File: Open the Lock/Open_the_Lock.cpp
// 经典的广度优先搜索

// Runtime: 132 ms, faster than 79.06% of C++ online submissions for Open the Lock.
// Memory Usage: 33.9 MB, less than 53.85% of C++ online submissions for Open the Lock.

class Solution 
{
public:
    int openLock(vector<string>& deadends, string target) 
    {
        // 初始状态为 0000
        string curState(4, '0');

        unordered_set<string> deadwords(deadends.begin(), deadends.end());
        unordered_set<string> hashset;

        queue<string> q;
        q.push(curState);
        hashset.insert(curState);

        int res = 0;
        while (!q.empty())
        {
            for (int size = q.size(); size > 0; size--)
            {
                curState = q.front();
                q.pop();

                if (curState == target)
                    return res;
                else if (deadwords.find(curState) != deadwords.end())
                    continue;
                else
                    candidate(q, hashset, curState); 
            }
            res++;
        }  
        return -1;
    }

    void candidate(queue<string>& q, unordered_set<string>& hashset, string& curState)
    {
        for (int i = 0; i < 4; i++)
        {
            char old = curState[i];
            
            curState[i] = curState[i] == '9' ? '0' : curState[i] + 1;
            if (hashset.find(curState) == hashset.end())
            {
                q.push(curState);
                hashset.insert(curState);
            }
            curState[i] = old;

            curState[i] = curState[i] == '0' ? '9' : curState[i] - 1;           
            if (hashset.find(curState) == hashset.end()) 
            {
                q.push(curState);
                hashset.insert(curState);
            }
            curState[i] = old;
        }
    }
};

File: Open the Lock/Open_the_Lock.py
# Runtime: 536 ms, faster than 78.17% of Python3 online submissions for Open the Lock.
# Memory Usage: 14.2 MB, less than 12.28% of Python3 online submissions for Open the Lock.

class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        
        curState = '0000'

        deadwords = set(deadends)
        hashset = set()

        queue = []
        queue.append(curState)
        hashset.add(curState)

        res = 0
        while queue:
            for _ in range(len(queue)):
                curState = queue.pop(0)
                if curState == target:
                    return res
                elif curState in deadwords:
                    continue
                else:
                    self.candidate(queue, hashset, curState)
            res += 1
        return -1
    
    def candidate(self, queue, hashset, curState):
        for i in range(4):
            temp = '0' if curState[i] == '9' else chr(ord(curState[i]) + 1)
            after = curState[:i] + temp + curState[i + 1:]

            temp = '9' if curState[i] == '0' else chr(ord(curState[i]) - 1)
            before = curState[:i] + temp + curState[i + 1:]

            if after not in hashset:
                queue.append(after)
                hashset.add(after)
            
            if before not in hashset:
                queue.append(before)
                hashset.add(before)